lib/routes/account.js

```diff
@@ -1251,49 +1404,116 @@ module.exports = function (
         var password = new Password(authPW, authSalt, config.verifierVersion)
-        return password.verifyHash()
-          .then(
-            function (verifyHash) {
-              return db.resetAccount(
-                accountResetToken,
-                {
-                  authSalt: authSalt,
-                  verifyHash: verifyHash,
-                  wrapWrapKb: crypto.randomBytes(32),
-                  verifierVersion: password.version
-                }
-              )
-            }
-          )
-          .then(
-            function () {
-              // Notify all devices that the account has changed.
-              push.notifyUpdate(accountResetToken.uid, 'passwordReset')
+        var account, sessionToken, keyFetchToken, verifyHash, wrapKb
 
-              return db.account(accountResetToken.uid)
-            }
-          )
-          .then(
-            function (account) {
-              log.activityEvent('account.reset', request, {
-                uid: account.uid.toString('hex')
+        return verifyAndResetAccount()
+          .then(createSessionToken)
+          .then(createKeyFetchToken)
+          .then(createResponse)
+          .done(reply, reply)
+
+        function verifyAndResetAccount () {
+          return password.verifyHash()
+            .then(
+              function (verifyHashData) {
+                verifyHash = verifyHashData
+
+                return db.resetAccount(
+                  accountResetToken,
+                  {
+                    authSalt: authSalt,
+                    verifyHash: verifyHash,
+                    wrapWrapKb: crypto.randomBytes(32),
+                    verifierVersion: password.version
+                  }
+                )
+              }
+            )
+            .then(
+              function () {
+                // Notify all devices that the account has changed.
+                push.notifyUpdate(accountResetToken.uid, 'passwordReset')
+
+                return db.account(accountResetToken.uid)
+              }
+            )
+            .then(
+              function (accountData) {
+                account = accountData
+                log.activityEvent('account.reset', request, {
+                  uid: account.uid.toString('hex')
+                })
+                log.event(
+                  'reset',
+                  {
+                    uid: account.uid.toString('hex') + '@' + config.domain,
+                    generation: account.verifierSetAt
+                  }
+                )
+                return customs.reset(account.email)
+              }
+            )
+            .then(
+              function () {
+                return password.unwrap(account.wrapWrapKb)
+              }
+            )
+            .then(
+              function (wrapKbData) {
+                wrapKb = wrapKbData
+              }
+            )
+        }
+
+        function createSessionToken () {
+          // Create a verified sessionToken
```

Comment:
Nit: please also comment about why it's ok to create it in the verified state here